In [20]:
import numpy as np
import pandas as pd
df = pd.read_csv('solana_data_mod.csv', sep='\t', index_col=0)

In [21]:
df

Timestamp       Price     MarketCap  TotalVolumes  TimestampMod  \
0    1680307200000   21.205402  8.194996e+09  4.159474e+08  1.680307e+09   
1    1680393600000   21.083500  8.133720e+09  2.519486e+08  1.680394e+09   
2    1680480000000   20.565439  7.929638e+09  2.509397e+08  1.680480e+09   
3    1680566400000   20.436690  7.900169e+09  5.297529e+08  1.680566e+09   
4    1680652800000   20.921101  8.093519e+09  3.484730e+08  1.680653e+09   
..             ...         ...           ...           ...           ...   
360  1711411200000  188.996855  8.383933e+10  4.384505e+09  1.711411e+09   
361  1711497600000  191.236145  8.483871e+10  4.010910e+09  1.711498e+09   
362  1711584000000  186.138944  8.236008e+10  4.361858e+09  1.711584e+09   
363  1711670400000  189.574991  8.424885e+10  3.248951e+09  1.711670e+09   
364  1711759271000  191.472923  8.516714e+10  3.344562e+09  1.711759e+09   

                    date  year  month  day  hour  minute  second  
0    2023-04-01 00:00:00  2023      4    1     0       0       0  
1    2023-04-02 00:00:00  2023      4    2     0       0       0  
2    2023-04-03 00:00:00  2023      4    3     0       0       0  
3    2023-04-04 00:00:00  2023      4    4     0       0       0  
4    2023-04-05 00:00:00  2023      4    5     0       0       0  
..                   ...   ...    ...  ...   ...     ...     ...  
360  2024-03-26 00:00:00  2024      3   26     0       0       0  
361  2024-03-27 00:00:00  2024      3   27     0       0       0  
362  2024-03-28 00:00:00  2024      3   28     0       0       0  
363  2024-03-29 00:00:00  2024      3   29     0       0       0  
364  2024-03-30 00:41:11  2024      3   30     0      41      11  

[365 rows x 12 columns]

In [22]:
df['State'] = np.where(df['Price'].diff() > 0, 'up', 'down')

In [23]:
df

Timestamp       Price     MarketCap  TotalVolumes  TimestampMod  \
0    1680307200000   21.205402  8.194996e+09  4.159474e+08  1.680307e+09   
1    1680393600000   21.083500  8.133720e+09  2.519486e+08  1.680394e+09   
2    1680480000000   20.565439  7.929638e+09  2.509397e+08  1.680480e+09   
3    1680566400000   20.436690  7.900169e+09  5.297529e+08  1.680566e+09   
4    1680652800000   20.921101  8.093519e+09  3.484730e+08  1.680653e+09   
..             ...         ...           ...           ...           ...   
360  1711411200000  188.996855  8.383933e+10  4.384505e+09  1.711411e+09   
361  1711497600000  191.236145  8.483871e+10  4.010910e+09  1.711498e+09   
362  1711584000000  186.138944  8.236008e+10  4.361858e+09  1.711584e+09   
363  1711670400000  189.574991  8.424885e+10  3.248951e+09  1.711670e+09   
364  1711759271000  191.472923  8.516714e+10  3.344562e+09  1.711759e+09   

                    date  year  month  day  hour  minute  second State  
0    2023-04-01 00:00:00  2023      4    1     0       0       0  down  
1    2023-04-02 00:00:00  2023      4    2     0       0       0  down  
2    2023-04-03 00:00:00  2023      4    3     0       0       0  down  
3    2023-04-04 00:00:00  2023      4    4     0       0       0  down  
4    2023-04-05 00:00:00  2023      4    5     0       0       0    up  
..                   ...   ...    ...  ...   ...     ...     ...   ...  
360  2024-03-26 00:00:00  2024      3   26     0       0       0    up  
361  2024-03-27 00:00:00  2024      3   27     0       0       0    up  
362  2024-03-28 00:00:00  2024      3   28     0       0       0  down  
363  2024-03-29 00:00:00  2024      3   29     0       0       0    up  
364  2024-03-30 00:41:11  2024      3   30     0      41      11    up  

[365 rows x 13 columns]

In [24]:
df['State'].unique()

array(['down', 'up'], dtype=object)

In [25]:
len(df)

365

In [26]:
# Total Number of Previous States that were up or down, which is just the count of all Up states or Down states
total_up = (df["State"] == "up").sum()
total_down = (df["State"] == "down").sum()
if (len(df["State"]) == total_up + total_down):
    print("Addition matches")
else:
    print(f"length of states: {len(df['State'])} <--> Added total_up + total_down: {total_up + total_down}")
# Number of times state was up and previous state was up
up_up_days = ((df["State"] == "up") & (df["State"].shift(1) == "up")).sum()

# Number of times state was up and previous state was down
up_down_days = ((df["State"] == "up") & (df["State"].shift(1) == "down")).sum()

# Number of times state was down and previous state was up
down_up_days = ((df["State"] == "down") & (df["State"].shift(1) == "up")).sum()

# Number of times state was down and previous state was down
down_down_days = ((df["State"] == "down") & (df["State"].shift(1) == "down")).sum()


print(f"Number of consecutive up-up days: {up_up_days}")
print(f"Number of consecutive up-down days: {up_down_days}")
print(f"Number of consecutive down-up days: {down_up_days}")
print(f"Number of consecutive down-down days: {down_down_days}")
print(f"Number of up states: {total_up}")
print(f"Number of up states: {total_down}")


Addition matches
Number of consecutive up-up days: 90
Number of consecutive up-down days: 99
Number of consecutive down-up days: 98
Number of consecutive down-down days: 77
Number of up states: 189
Number of up states: 176


In [ ]:
# say that we want to know do the same for different states 
# lets say you want to know when the equity was up -> up -> ?
'''
P(up-up-up), P(up-up-down)
P(up-down-up), P(up-down-down)
P(down-up-up), P(down-up-down)
P(down-down-up), P(down-down-down)

Remember that we have a larger matrix now because we increased n by 1 so 2^n where n=3 ==> 8

and keep in mind we need total amount we find up-up-up / total amount of times we count up-up for instance P(up-up-up)
'''

In [27]:
# create transition matrix
up_up_ratio = round(up_up_days/total_up, 2)
up_down_ratio = round(up_down_days/total_up, 2)
down_up_ratio = round(down_up_days/total_down, 2)
down_down_ratio = round(down_down_days/total_down, 2)

print(f"up-up ratio: {up_up_ratio}")
print(f"up-down ratio: {up_down_ratio}")
print(f"down-up ratio: {down_up_ratio}")
print(f"down-down ratio: {down_down_ratio}")


up-up ratio: 0.48
up-down ratio: 0.52
down-up ratio: 0.56
down-down ratio: 0.44


In [28]:
transition_matrix = pd.DataFrame(
    {"up": [up_up_ratio, down_up_ratio] 
     ,"down": [up_down_ratio, down_down_ratio]}, index = ["up", "down"])
# index are the rows... which can sometimes be confusing because it looks like it's in row format when you code it,
# but it is actually in column format.  think of it was "up" : [x, y] as a column vector

In [29]:
print(transition_matrix)

        up  down
up    0.48  0.52
down  0.56  0.44


In [30]:
# Lets do it the silly way first 

P = transition_matrix.to_numpy()
print("Transition Matrix P = \n",P)

one = np.linalg.matrix_power(P, 1)
two = np.linalg.matrix_power(P, 2)
five = np.linalg.matrix_power(P, 5)
twenty = np.linalg.matrix_power(P, 20)
one_hundo = np.linalg.matrix_power(P, 100)

pi = np.array([0,1]) # this means it was a down day in previous state

print(f"power of 1:\n  {one}")
print(f"power of 2:\n  {two}")
print(f"power of 5:\n  {five}")
print(f"power of 20:\n  {twenty}")
print(f"power of 100:\n  {one_hundo}")


Transition Matrix P = 
 [[0.48 0.52]
 [0.56 0.44]]
power of 1:
  [[0.48 0.52]
 [0.56 0.44]]
power of 2:
  [[0.5216 0.4784]
 [0.5152 0.4848]]
power of 5:
  [[0.51851694 0.48148306]
 [0.51852022 0.48147978]]
power of 20:
  [[0.51851852 0.48148148]
 [0.51851852 0.48148148]]
power of 100:
  [[0.51851852 0.48148148]
 [0.51851852 0.48148148]]


In [33]:
import scipy.linalg
values, left_vectors = scipy.linalg.eig(P, right=False, left=True)

print("EigenValues: ", values)
print("EigneVectors:\n ", left_vectors)


EigenValues:  [ 1.  +0.j -0.08+0.j]
EigneVectors:
  [[ 0.73279349 -0.70710678]
 [ 0.6804511   0.70710678]]


In [32]:
pi = left_vectors[:,0]
pi_normalized = [(x/np.sum(pi)).real for x in pi]
pi_normalized


[0.5185185185185185, 0.48148148148148145]

So now you might be asking... well what do these numbers mean.... well thats a great question

It means that in any random walk, wherever we start, we end with these two probabilities.

So if we start at "UP" [1,0] we take a random walk, our equilibrium lambda = 1, will give you those probabilies. from any initial state if up next day we have ~52% chance it will be up and ~48% itll be down